File: apps/api/src/schemas.py
from __future__ import annotations

from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, field_validator

from src.models import (
    ClipReframeMode,
    ClipRenderType,
    ClipSource,
    ClipStatus,
    SermonStatus,
)


class StrictBaseModel(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)


class ORMModel(StrictBaseModel):
    model_config = ConfigDict(extra="forbid", strict=True, from_attributes=True)


class SermonPayload(StrictBaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    preacher: Optional[str] = None
    series: Optional[str] = None
    sermon_date: Optional[date] = None
    tags: Optional[list[str]] = None

    @field_validator("sermon_date", mode="before")
    @classmethod
    def parse_sermon_date(cls, value):
        if value in (None, ""):
            return None
        if isinstance(value, date):
            return value
        if isinstance(value, str):
            try:
                return date.fromisoformat(value)
            except ValueError as exc:
                raise ValueError("Invalid date format; expected YYYY-MM-DD") from exc
        return value


class SermonCreate(SermonPayload):
    filename: Optional[str] = None


class SermonUpdate(SermonPayload):
    source_url: Optional[str] = None
    status: Optional[SermonStatus] = None


class SermonCreateResponse(StrictBaseModel):
    sermon: SermonRead
    upload_url: str
    object_key: str


class UploadCompleteResponse(StrictBaseModel):
    sermon: SermonRead


class SuggestClipsResponse(StrictBaseModel):
    sermon_id: int
    status: str


class EmbedResponse(StrictBaseModel):
    sermon_id: int
    status: str


class SermonRead(ORMModel):
    id: int
    title: Optional[str]
    description: Optional[str] = None
    preacher: Optional[str] = None
    series: Optional[str] = None
    sermon_date: Optional[date] = None
    tags: Optional[list[str]] = None
    source_url: Optional[str]
    source_download_url: Optional[str] = None
    progress: int
    status: SermonStatus
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None


class ClipCreate(StrictBaseModel):
    sermon_id: int
    start_ms: int
    end_ms: int
    source: Optional[ClipSource] = None
    score: Optional[float] = None
    rationale: Optional[str] = None
    template_id: Optional[str] = None
    reframe_mode: Optional[ClipReframeMode] = None
    render_type: Optional[ClipRenderType] = None


class ClipUpdate(StrictBaseModel):
    start_ms: Optional[int] = None
    end_ms: Optional[int] = None
    status: Optional[ClipStatus] = None
    output_url: Optional[str] = None
    source: Optional[ClipSource] = None
    score: Optional[float] = None
    rationale: Optional[str] = None
    llm_trim: Optional[dict] = None
    llm_trim_confidence: Optional[float] = None
    trim_applied: Optional[bool] = None
    template_id: Optional[str] = None
    reframe_mode: Optional[ClipReframeMode] = None
    render_type: Optional[ClipRenderType] = None


class ClipRead(ORMModel):
    id: int
    sermon_id: int
    start_ms: int
    end_ms: int
    output_url: Optional[str]
    download_url: Optional[str] = None
    status: ClipStatus
    source: ClipSource
    score: Optional[float]
    rationale: Optional[str]
    use_llm: bool = False
    llm_prompt_tokens: Optional[int] = None
    llm_completion_tokens: Optional[int] = None
    llm_total_tokens: Optional[int] = None
    llm_estimated_cost: Optional[float] = None
    llm_output_tokens: Optional[int] = None
    llm_cache_hit_tokens: Optional[int] = None
    llm_cache_miss_tokens: Optional[int] = None
    llm_method: Optional[str] = None
    llm_trim: Optional[dict] = None
    llm_trim_confidence: Optional[float] = None
    trim_applied: bool = False
    template_id: Optional[str]
    reframe_mode: ClipReframeMode
    render_type: ClipRenderType
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None


class ClipSuggestionsResponse(StrictBaseModel):
    sermon_id: int
    clips: list[ClipRead]


class ClipAcceptResponse(StrictBaseModel):
    suggestion_id: int
    clip: ClipRead


class ClipFeedbackCreate(StrictBaseModel):
    accepted: bool
    user_id: Optional[str] = None


class ClipFeedbackRead(ORMModel):
    id: int
    clip_id: int
    accepted: bool
    user_id: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None


class ClipRenderResponse(StrictBaseModel):
    clip_id: int
    status: str
    render_type: ClipRenderType


class TranscriptSegmentRead(ORMModel):
    id: int
    sermon_id: int
    start_ms: int
    end_ms: int
    text: str
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None


class SearchResult(StrictBaseModel):
    segment_id: int
    text: str
    start_ms: int
    end_ms: int


class SearchResponse(StrictBaseModel):
    sermon_id: int
    query: str
    results: list[SearchResult]


class TemplateConfig(StrictBaseModel):
    font: str
    font_size: int
    y_pos: int
    max_words_per_line: int
    highlight_mode: str
    safe_margins: dict


class TemplateCreate(StrictBaseModel):
    id: str
    name: str
    config_json: TemplateConfig


class TemplateRead(ORMModel):
    id: str
    name: str
    config_json: TemplateConfig
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None

File: apps/web/lib/schemas.js
import { z } from "zod";

export const sermonStatusSchema = z.enum([
  "pending",
  "uploaded",
  "processing",
  "transcribed",
  "suggested",
  "embedded",
  "error",
  "completed",
  "failed"
]);

export const clipStatusSchema = z.enum(["pending", "processing", "done", "error"]);
export const clipSourceSchema = z.enum(["manual", "auto"]);
export const clipReframeModeSchema = z.enum(["center", "face"]);
export const clipRenderTypeSchema = z.enum(["preview", "final"]);

export const sermonSchema = z.object({
  id: z.number(),
  title: z.string().nullable(),
  description: z.string().nullable().optional(),
  preacher: z.string().nullable().optional(),
  series: z.string().nullable().optional(),
  sermon_date: z.string().nullable().optional(),
  tags: z.array(z.string()).nullable().optional(),
  source_url: z.string().nullable(),
  source_download_url: z.string().nullable().optional(),
  progress: z.number(),
  status: sermonStatusSchema,
  error_message: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
  deleted_at: z.string().nullable().optional()
});

export const sermonListSchema = z.array(sermonSchema);

export const clipSchema = z.object({
  id: z.number(),
  sermon_id: z.number(),
  start_ms: z.number(),
  end_ms: z.number(),
  output_url: z.string().nullable(),
  download_url: z.string().nullable().optional(),
  status: clipStatusSchema,
  source: clipSourceSchema,
  score: z.number().nullable().optional(),
  rationale: z.string().nullable().optional(),
  use_llm: z.boolean(),
  llm_prompt_tokens: z.number().nullable().optional(),
  llm_completion_tokens: z.number().nullable().optional(),
  llm_total_tokens: z.number().nullable().optional(),
  llm_estimated_cost: z.number().nullable().optional(),
  llm_output_tokens: z.number().nullable().optional(),
  llm_cache_hit_tokens: z.number().nullable().optional(),
  llm_cache_miss_tokens: z.number().nullable().optional(),
  llm_method: z.string().nullable().optional(),
  llm_trim: z.record(z.any()).nullable().optional(),
  llm_trim_confidence: z.number().nullable().optional(),
  trim_applied: z.boolean(),
  template_id: z.string().nullable().optional(),
  reframe_mode: clipReframeModeSchema,
  render_type: clipRenderTypeSchema,
  created_at: z.string(),
  updated_at: z.string(),
  deleted_at: z.string().nullable().optional()
});

export const clipListSchema = z.array(clipSchema);

export const transcriptSegmentSchema = z.object({
  id: z.number(),
  sermon_id: z.number(),
  start_ms: z.number(),
  end_ms: z.number(),
  text: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
  deleted_at: z.string().nullable().optional()
});

export const transcriptSegmentListSchema = z.array(transcriptSegmentSchema);

export const createSermonResponseSchema = z.object({
  sermon: sermonSchema,
  upload_url: z.string(),
  object_key: z.string()
});

export const uploadCompleteResponseSchema = z.object({
  sermon: sermonSchema
});

export const embedResponseSchema = z.object({
  sermon_id: z.number(),
  status: z.string()
});

export const suggestClipsResponseSchema = z.object({
  sermon_id: z.number(),
  status: z.string()
});

export const clipSuggestionsResponseSchema = z.object({
  sermon_id: z.number(),
  clips: z.array(clipSchema)
});

export const clipAcceptResponseSchema = z.object({
  suggestion_id: z.number(),
  clip: clipSchema
});

export const clipFeedbackSchema = z.object({
  id: z.number(),
  clip_id: z.number(),
  accepted: z.boolean(),
  user_id: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
  deleted_at: z.string().nullable().optional()
});

export const clipRenderResponseSchema = z.object({
  clip_id: z.number(),
  status: z.string(),
  render_type: clipRenderTypeSchema
});

export const searchResultSchema = z.object({
  segment_id: z.number(),
  text: z.string(),
  start_ms: z.number(),
  end_ms: z.number()
});

export const searchResponseSchema = z.object({
  sermon_id: z.number(),
  query: z.string(),
  results: z.array(searchResultSchema)
});
